src/mlpack/core/data/split_data.hpp

```diff
@@ -14,9 +14,139 @@
+                     const arma::Row<U>& inputLabel,
+                     arma::Mat<T>& trainData,
+                     arma::Mat<T>& testData,
+                     arma::Row<U>& trainLabel,
+                     arma::Row<U>& testLabel,
+                     const double testRatio,
+                     const bool shuffleData = true)
+{
+  /**
+   * Basic idea:
+   * Let us say we have to stratify a dataset based on labels:
+   * 0 0 0 0 0 1 1 1 1 1 1 (5 0s, 6 1s)
+   *
+   * Let our test ratio be 0.2:
+   * We visit each label and keep the count of each label in our unordered map
+   *
+   * Whenever we encounter a label, we calculate the floored values of
+   * current_count * test_ratio --- labelMap[label] * testRatio and
+   * current_count+1 * test_ratio --- (labelMap[label] + 1) * testRatio
+   *
+   * If there is a difference in counts, then we add to our test set
+   *
+   * Considering our example
+   * 0 -- train set ( 0 * 0.2 == 1 * 0.2 ) (After floor)
+   * 0 -- train set ( 1 * 0.2 == 2 * 0.2 ) (After floor)
+   * 0 -- train set ( 2 * 0.2 == 3 * 0.2 ) (After floor)
+   * 0 -- train set ( 3 * 0.2 == 4 * 0.2 ) (After floor)
+   * 0 -- test set  ( 4 * 0.2 <  5 * 0.2 ) (After floor)
+   *
+   * 1 -- train set ( 0 * 0.2 == 1 * 0.2 ) (After floor)
+   * 1 -- train set ( 1 * 0.2 == 2 * 0.2 ) (After floor)
+   * 1 -- train set ( 2 * 0.2 == 3 * 0.2 ) (After floor)
+   * 1 -- train set ( 3 * 0.2 == 4 * 0.2 ) (After floor)
+   * 1 -- test set  ( 4 * 0.2 <  5 * 0.2 ) (After floor)
+   * 1 -- train set ( 5 * 0.2 == 6 * 0.2 ) (After floor)
+   *
+   * Finally
+   * train set,
+   * 0 0 0 0 1 1 1 1 1 (4 0s, 5 1s)
+   *
+   * test set,
+   * 0 1
+   */
+  arma::uvec indices;
+  indices.set_size(inputLabel.n_cols);
+
+  size_t trainIdx = 0;
+  size_t testIdx = inputLabel.n_cols - 1;
+  std::unordered_map<U, size_t> labelMap;
+
+  arma::uvec order =
+      arma::linspace<arma::uvec>(0, input.n_cols - 1, input.n_cols);
+
+  if (shuffleData)
+  {
+    order = arma::shuffle(order);
+  }
+
+  for (auto i: order)
+  {
+    auto label = inputLabel[i];
+
+    if (floor(labelMap[label] * testRatio) ==
+        floor((labelMap[label] + 1) * testRatio))
+    {
+      indices[trainIdx] = i;
+      trainIdx += 1;
+    }
+    else
+    {
+      indices[testIdx] = i;
+      testIdx -= 1;
+    }
+    labelMap[label] += 1;
+  }
+
+  const size_t testSize = indices.n_rows - trainIdx;
+  const size_t trainSize = indices.n_rows - testSize;
+  trainData.set_size(input.n_rows, trainSize);
+  testData.set_size(input.n_rows, testSize);
+  trainLabel.set_size(trainSize);
+  testLabel.set_size(testSize);
+
+  if (trainSize > 0)
+  {
+    trainData = input.cols(indices.subvec(0, trainIdx - 1));
+    trainLabel = inputLabel.cols(indices.subvec(0, trainIdx - 1));
+  }
+  if (trainSize < input.n_cols)
+  {
+    testData = input.cols(indices.subvec(trainIdx, indices.n_rows - 1));
+    testLabel = inputLabel.cols(indices.subvec(trainIdx, indices.n_rows - 1));
+  }
```

Comment:
Actually this ends up taking four passes over the data! :smile:  This is because each call to `cols()` to extract a new matrix will take an entire pass over `input`.

If you wanted to make it more efficient, you could write a single loop over `input` that copies each column at a time.  However, ideally we want to pass over `input` and `inputLabel` contiguously (i.e. access the first column, then the second column, etc...) for the sake of speed.  So in order to do that, the strategy I suggested above to avoid using `indices` might be necessary.

But, in any case, you could time both ways.  I'm pretty sure that the strategy I proposed could be faster but I'm not absolutely certain. :+1:

Reply:
Completely slipped my mind about the passes in cols!. 😆 
Thanks for pointing that out. 

Removing the ```indices``` vector makes sense. Will experiment and get back on this.